File: java/Dson-Codec/src/main/java/cn/wjybxx/dsoncodec/codecs/ObjectLitePtrCodec.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dsoncodec.codecs;

import cn.wjybxx.dson.text.ObjectStyle;
import cn.wjybxx.dson.types.ObjectLitePtr;
import cn.wjybxx.dsoncodec.DsonCodec;
import cn.wjybxx.dsoncodec.DsonObjectReader;
import cn.wjybxx.dsoncodec.DsonObjectWriter;
import cn.wjybxx.dsoncodec.TypeInfo;
import cn.wjybxx.dsoncodec.annotations.DsonCodecScanIgnore;

import javax.annotation.Nonnull;
import java.util.function.Supplier;

/**
 * @author wjybxx
 * date - 2024/1/7
 */
@DsonCodecScanIgnore
public class ObjectLitePtrCodec implements DsonCodec<ObjectLitePtr> {

    @Nonnull
    @Override
    public TypeInfo getEncoderType() {
        return TypeInfo.of(ObjectLitePtr.class);
    }

    @Override
    public boolean autoStartEnd() {
        return false;
    }

    @Override
    public void writeObject(DsonObjectWriter writer, ObjectLitePtr inst, TypeInfo declaredType, ObjectStyle style) {
        writer.writeLitePtr(null, inst);
    }

    @Override
    public ObjectLitePtr readObject(DsonObjectReader reader, Supplier<? extends ObjectLitePtr> factory) {
        return reader.readLitePtr(null);
    }
}

File: csharp/Wjybxx.Dson.Codec/src/Codecs/ObjectLitePtrCodec.cs
#region LICENSE

// Copyright 2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using Wjybxx.Dson.Text;
using Wjybxx.Dson.Types;

namespace Wjybxx.Dson.Codec.Codecs
{
public class ObjectLitePtrCodec : IDsonCodec<ObjectLitePtr>
{
    public bool AutoStartEnd => false;

    public void WriteObject(IDsonObjectWriter writer, ref ObjectLitePtr inst, Type declaredType, ObjectStyle style) {
        writer.WriteLitePtr(null, in inst);
    }

    public ObjectLitePtr ReadObject(IDsonObjectReader reader, Type declaredType, Func<ObjectLitePtr>? factory = null) {
        return reader.ReadLitePtr(reader.CurrentName);
    }
}
}